examples/md-flexible/src/parsing/MDFlexConfig.h

```diff
@@ -0,0 +1,476 @@
+ */
+class MDFlexConfig {
+ public:
+  MDFlexConfig() = default;
+
+  /**
+   * Struct to bundle information for options.
+   * @tparam T Datatype of the option
+   * @tparam getOptChar Char for the switch case that is used during cli argument parsing with getOpt.
+   * @note Set to 0 (not '0') if not intended to be used from command line.
+   *
+   */
+  template <class T, char getOptChar>
+  struct MDFlexOption {
+    /**
+     * Value of this option.
+     */
+    T value;
+
+    /**
+     * Indicate whether this option is a flag or takes arguments.
+     */
+    bool requiresArgument;
+
+    /**
+     * String representation of the option name.
+     */
+    std::string name;
+
+    /**
+     * String describing this option. This is displayed when md-flexible is invoked with --help.
+     */
+    std::string description;
+
+    /**
+     * Member to access the template parameter.
+     */
+    constexpr static char getoptChar{getOptChar};
+
+    /**
+     * Constructor
+     * @param value Default value for this option.
+     * @param newName String representation of the option name.
+     * @param requiresArgument Indicate whether this option is a flag or takes arguments.
+     * @param newDescription String describing this option. This is displayed when md-flexible is invoked with --help.
+     */
+    MDFlexOption(T value, std::string newName, bool requiresArgument, std::string newDescription)
+        : requiresArgument(requiresArgument),
+          name(std::move(newName)),
+          description(std::move(newDescription)),
+          value(std::move(value)) {}
+
+    /**
+     * Returns a getopt option struct for this object.
+     * @return
+     */
+    [[nodiscard]] struct option toGetoptOption() const {
+      struct option retStruct {
+        name.c_str(), requiresArgument, nullptr, getOptChar
+      };
+      return retStruct;
+    }
+  };
+
+  /**
+   * Convert the content of the config to a string representation.
+   * @return
+   */
+  [[nodiscard]] std::string to_string() const;
+
+  /**
+   * Checks parsed Objects and determines the necessary size of the simulation box.
+   */
+  void calcSimulationBox();
+
+  /**
+   * Adds parameters to all relevant particle property attributes and checks if the type already exists.
+   * @param typeId
+   * @param epsilon
+   * @param sigma
+   * @param mass
+   */
+  void addParticleType(unsigned long typeId, double epsilon, double sigma, double mass);
+
+  /**
+   * Choice of the functor
+   */
+  enum class FunctorOption { lj12_6, lj12_6_AVX, lj12_6_Globals };
+
+  /**
+   * Choice of the particle generators specified in the command line
+   */
+  enum class GeneratorOption { grid, uniform, gaussian, sphere };
+
+  //  All options in the config
+  //  Make sure that the description is parsable by `CLIParser::createZSHCompletionFile()`!
+
+  /**
+   * yamlFilename
+   */
+  MDFlexOption<std::string, 'Y'> yamlFilename{"", "yaml-filename", true, "Path to input file."};
```

Comment:
It now would be totally awesome if we could drop specifying the getoptChar (here 'Y') and generate it from e.g. the name (here "yaml-filename"). For this to work, however, the name needs to be available at compile-time, because the getoptChar needs to be available at compile time. [This Stackoverflow post](https://stackoverflow.com/a/1826505/7019073) seems to suggest that it is possible but I do not see how this can be done here easily.

Reply:
This will be supported in c++20 via [P0732R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0732r2.pdf) "Class types in non-type template parameters". This is in gcc9 and msvc, but not yet in clang.

Reply:
(at least partially)

Reply:
for now, I'd suggest to use `__LINE__`.

Reply:
done with 8e93ac8